<a href="https://colab.research.google.com/github/bachecodes/data-science-bootcamp-selection-projects/blob/main/pandas/pandas_basics.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Part 1: Create Your Own Dataset (Foundation)
Task: Build a DataFrame from a Dictionary

Use a dictionary to create a dataset of your choice.

    It needs to have 5 Columns (features)
    It needs to have 15 rows
    It needs to assign custom index.


In [1]:
import pandas as pd

# Create dictionary
data = {
    "Name": ["Ali", "Sara", "John", "Marta", "David", "Lina", "Tom", "Nina", "Sam", "Eva", "Leo", "Maya", "Noah", "Zara", "Adam"],
    "Age": [23, 25, 30, 22, 35, 28, 40, 19, 33, 27, 29, 24, 31, 26, 38],
    "Gender": ["M", "F", "M", "F", "M", "F", "M", "F", "M", "F", "M", "F", "M", "F", "M"],
    "Score": [88, 92, 85, 90, 78, 95, 70, 89, 82, 91, 87, 93, 80, 94, 76],
    "Passed": ["Yes", "Yes", "Yes", "Yes", "No", "Yes", "No", "Yes", "Yes", "Yes", "Yes", "Yes", "Yes", "Yes", "No"]
}

# Custom index
index_labels = [f"S{i}" for i in range(1, 16)]

df_custom = pd.DataFrame(data, index=index_labels)

print(df_custom)

      Name  Age Gender  Score Passed
S1     Ali   23      M     88    Yes
S2    Sara   25      F     92    Yes
S3    John   30      M     85    Yes
S4   Marta   22      F     90    Yes
S5   David   35      M     78     No
S6    Lina   28      F     95    Yes
S7     Tom   40      M     70     No
S8    Nina   19      F     89    Yes
S9     Sam   33      M     82    Yes
S10    Eva   27      F     91    Yes
S11    Leo   29      M     87    Yes
S12   Maya   24      F     93    Yes
S13   Noah   31      M     80    Yes
S14   Zara   26      F     94    Yes
S15   Adam   38      M     76     No


# Part 2: Real Dataset — Titanic
Use: Titanic Dataset (train.csv)



In [11]:
path= "/content/drive/MyDrive/Titanic-Dataset.csv"

In [12]:
df = pd.read_csv(path)

# Step 1: Exploration

    .head()
    .info()
    .describe()


In [13]:
df.head()


,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S


In [14]:
print(df.info())


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB
None


In [15]:
df.describe()

,PassengerId,Survived,Pclass,Age,SibSp,Parch,Fare
count,891.000000,891.000000,891.000000,714.000000,891.000000,891.000000,891.000000
mean,446.000000,0.383838,2.308642,29.699118,0.523008,0.381594,32.204208
std,257.353842,0.486592,0.836071,14.526497,1.102743,0.806057,49.693429
min,1.000000,0.000000,1.000000,0.420000,0.000000,0.000000,0.000000
25%,223.500000,0.000000,2.000000,20.125000,0.000000,0.000000,7.910400
50%,446.000000,0.000000,3.000000,28.000000,0.000000,0.000000,14.454200
75%,668.500000,1.000000,3.000000,38.000000,1.000000,0.000000,31.000000
max,891.000000,1.000000,3.000000,80.000000,8.000000,6.000000,512.329200


# Step 2: Data Cleaning
Handle missing values:

    Age → median
    Embarked → mode
    Cabin → drop
    Remove duplicates if any

In [44]:
df.isnull().sum()

,0
PassengerId,0
Survived,0
Pclass,0
Name,0
Sex,0
Age,0
SibSp,0
Parch,0
Ticket,0
Fare,0


In [17]:
missing_percent = (df.isnull().sum() / len(df)) * 100
print(missing_percent)

PassengerId     0.000000
Survived        0.000000
Pclass          0.000000
Name            0.000000
Sex             0.000000
Age            19.865320
SibSp           0.000000
Parch           0.000000
Ticket          0.000000
Fare            0.000000
Cabin          77.104377
Embarked        0.224467
dtype: float64


from this we can understand that  2 columns age , and cabin have missing values
    
    age= 177 missing (19.865320)
    cabin= 687 missing (77.104377)
    Embarked  = 2 missing  0.224467
we have to use difffrent method for both   

age column have around 20 % missing so we can fill it with mean or meadins

but cabine column have missing greater than 70 % its better to drop this column


In [18]:
# Fill missing Age with median
df["Age"].fillna(df["Age"].median(), inplace=True)



/tmp/ipykernel_3184/3928094412.py:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df["Age"].fillna(df["Age"].median(), inplace=True)


In [19]:
# Fill missing Embarked with mode
df["Embarked"].fillna(df["Embarked"].mode()[0], inplace=True)



/tmp/ipykernel_3184/1306108366.py:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df["Embarked"].fillna(df["Embarked"].mode()[0], inplace=True)


In [20]:
# Drop Cabin column
df.drop(columns=["Cabin"], inplace=True)



In [21]:
# Remove duplicates
df.drop_duplicates(inplace=True)

we have to confirme that if there is missing value

In [22]:
df.isnull().sum()

,0
PassengerId,0
Survived,0
Pclass,0
Name,0
Sex,0
Age,0
SibSp,0
Parch,0
Ticket,0
Fare,0


# Step 3: Data Analysis
Using groupby():

    Survival rate by gender
    Survival rate by class
    Average age per class
    Survival rate by age group


In [23]:
survival_gender = df.groupby("Sex")["Survived"].mean()
print(survival_gender)

Sex
female    0.742038
male      0.188908
Name: Survived, dtype: float64


In [31]:
survival_gender = df.groupby("Sex")["Survived"].describe()
print(survival_gender)


        count      mean       std  min  25%  50%  75%  max
Sex                                                       
female  314.0  0.742038  0.438211  0.0  0.0  1.0  1.0  1.0
male    577.0  0.188908  0.391775  0.0  0.0  0.0  0.0  1.0


In [32]:
survival_class = df.groupby("Pclass")["Survived"].mean()
print(survival_class)

Pclass
1    0.629630
2    0.472826
3    0.242363
Name: Survived, dtype: float64


In [33]:
avg_age_class = df.groupby("Pclass")["Age"].mean()
print(avg_age_class)

Pclass
1    36.812130
2    29.765380
3    25.932627
Name: Age, dtype: float64


In [34]:
# Create age groups
df["AgeGroup"] = pd.cut(df["Age"],
                       bins=[0, 12, 18, 35, 60, 100],
                       labels=["Child", "Teen", "Young Adult", "Adult", "Senior"])

survival_age_group = df.groupby("AgeGroup")["Survived"].mean()
print(survival_age_group)

AgeGroup
Child          0.579710
Teen           0.428571
Young Adult    0.353271
Adult          0.400000
Senior         0.227273
Name: Survived, dtype: float64


/tmp/ipykernel_3184/3108116729.py:6: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  survival_age_group = df.groupby("AgeGroup")["Survived"].mean()


this implies kids in age group of 0-12 have high survival rate

# Step 4: Filtering
    Female passengers who survived
    Children who survived
    Passengers in 1st class with high survival probability


In [43]:
female_survived = df[(df["Sex"] == "female") & (df["Survived"] == 1)]
female_survived.PassengerId.count()

np.int64(233)

In [42]:
children_survived = df[(df["Age"] < 12) & (df["Survived"] == 1)]
children_survived.Age.value_counts()

,count
Age,
4.00,7
3.00,5
1.00,5
5.00,4
2.00,3
9.00,2
0.83,2
8.00,2
0.75,2


In [41]:
first_class_survived = df[(df["Pclass"] == 1) & (df["Survived"] == 1)]
first_class_survived.PassengerId.count()

np.int64(136)

# Step 5: Insights (VERY IMPORTANT)
Answer clearly:

    Who was more likely to survive? (gender)
    Did class affect survival?
    Were children prioritized?
    What combination had the highest survival rate?

In [45]:
df.groupby(["Sex", "Pclass"])["Survived"].mean()

Sex     Pclass
female  1         0.968085
        2         0.921053
        3         0.500000
male    1         0.368852
        2         0.157407
        3         0.135447
Name: Survived, dtype: float64

# Who was more likely to survive?

from the above dat we got :

    Female → ~0.74 (74%)
    Male → ~0.19 (19%)

this means
Out of every 10 women, about 7 survived, while only about 2 out of 10 men survived.

So this wasn’t a small difference — it was huge.
It clearly shows that during the chaos, people followed the *“Save the women first.”* rule

# Did class affect survival?

from the above data we got

    1st class → ~63%
    2nd class → ~47%
    3rd class → ~24%

example , from  10 passengers

In 1st class, about 6 survived
In 3rd class, only about 2 survived

this is a big gap.
People in 1st class had better access to lifeboats and help.

 Survival wasn’t just about luck the position on the ship mattered a lot

like :

If you were rich and in 1st class, you were closer to lifeboats and got help faster.
If you were in 3rd class, you were deeper in the ship and had fewer chances to escape.

So, survival wasn’t just luck — your social class played a big role.

# Were children prioritized?


Yes, children were clearly given priority.

    Child          0.579710(0-12)
    Teen           0.428571(12, 18)
    Young Adult    0.353271(18, 35)
    Adult          0.400000 (35, 60)
    Senior         0.227273 (60, 100)
Many children survived compared to adults, showing that people tried to protect the youngest passengers during the disaster.

It reflects a natural human instinct: save the kids first.

# What combination had the highest survival rate?


The people most likely to survive were:
This is where everything comes together:

A woman in 1st class was almost guaranteed to survive
A man in 3rd class had very little chance
Younger (especially children or young adults)


So if someone was a young woman traveling in 1st class, they had the best chances of making it out safely